docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+out a service-oriented architecture, which builds on platform-specific
+primitives for process management, IPC, and credentials management. Service-
+(and thus: network-) boundaries shall be defined mainly according to the
+lifecycle of the application.
+
+Note that this matters only for a usage scenario where all installed
+`radicle-link` applications share the same user-scoped state. An alternative
+strategy would be to scope all state to the application, ie. any two
+applications would not share the same storage nor _device key_. Since this
+raises the question of how those different states are synchronised in order to
+provide a coherent view to the user, we haven't explored this approach further,
+but admit that some clever engineering on the storage layer could enable it in
+the future.
+
+We dismiss the approach of providing a local monolithic server which encompasses
+all required functionality for modularity and extensibility reasons.
+
+
+## Prerequisites
+
+### Platforms
+
+For this discussion, we divide platform targets into tiers:
+
+* **Tier 1**:
+
+  > Fully supported platform, receives testing, deployment, and system
+  > configurations (may include packaging)
+
+  * Linux (`x86_64-unknown-linux-gnu`)
+  * macOS (`x86_64-apple-darwin`)
+
+* **Tier 2**:
+
+  > Basic support ("it compiles")
+
+  * Windows
+  * Other processor architectures for tier 1 platforms
+
+* **Tier 3**:
+
+  > Not supported, but potentially worth considering when assessing architecture
+  > decisions
+
+  * Mobile (iOS, Android)
+  * Wasm
+
+
+### System Services
+
+We will rely on declarative service / process management to be provided by the
+platform, specifically "socket activation".
+
+On macOS, this is provided on macOS via [launchd]. Linux distributions
+considered as tier 1 shall, however, by limited to those based on  [systemd] for
+the time being.
+
+We also assume the presence of an agent capable of producing Ed25519 signatures
+utilising the [SSH agent protocol][ssh-agent].
+
+On macOS, the distribution `ssh-agent` integrates with the system keychain.
+Other choices include the `ssh-agent` provided by the `openssh` suite, the
+`gpg-agent`, or (on Linux systems) `gnome-keyring`.
+
+Lastly, we consider [D-Bus][dbus] for publish-subscribe messaging. D-Bus is
+available on `systemd`-based systems per default, while it requires use of an
+external package manager on macOS.
+
+> Note: a D-Bus implementation is provided on `systemd`-based platforms, and can
+> be installed on macOS using `homebrew` or `MacPorts`. It may, however, turn
+> out to complicate builds due to varying link-time requirements. As we only
+> intend to use D-Bus for pub-sub messaging ("signals"), we may instead consider
+> to provide a purpose-built in-memory pub-sub service.
+
+
+### Signatures
+
+Each process which needs to produce signatures using the `radicle-link` _device
+key_, SHALL delegate this to an agent conforming to the [SSH agent
+protocol][ssh-agent] (section 4.5.).
```

Comment:
Could you expand on how this would work in practice? How would applications add the device key to the agent? How would they know which key is the device key? Do we want to make use of [key confirmation constraints](https://datatracker.ietf.org/doc/html/draft-miller-ssh-agent-02#section-4.2.6.2)?

One thing I’m concerned about is that if one service loads the key into the agent it is available to all services running as the user. This would make RPC request signing for process redundant from a security standpoint

Reply:
We’d provide a small tool to substitute `ssh-keygen`, which prompts for a passphrase, stores the key somewhere, and then imports it into `ssh-agent` (the protocol supports that). The gpg agent requires more configuration (of course), but I guess that’s for advanced users anyway.

Requesting a signature from the agent requires the public key and the payload to sign as parameters — so the application needs to store the public key somewhere.

Reply:
In the email notification you mention key confirmation constraints: I couldn’t really get those to work using `thrussh`, so that would require some experimentation. Also note that other agents which happen to support the agent protocol may not respect those at all (eg gpg agent only accepts manual configuration).

Ideally, the configuration should be such that not _every_ signature request requires pin entry — otherwise one would enter it twice for every command of authentication is in place. Idk... “requires further research”

Reply:
9540500a167b0d76fe26735882cc0ea9de0258c8